app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,157 @@
+      end
 
-  def batch_callbacks(batch)
-    %i[success complete].each { |name| batch.on(name, self.class, callback_options) }
-    yield
-    bid = batch.bid
+      ar_object_str = "#{ar_object.class}-#{ar_object.id}"
 
-    if Sidekiq::Batch::Status.new(bid).total.zero?
-      on_complete(bid, callback_options)
-    else
-      info("DeleteObjectHierarchyWorker#batch_success_callback retry job with the hierarchy of workers: #{caller_worker_hierarchy}")
-      retry_job wait: 5.minutes
+      [
+        "Plain-#{ar_object_str}",
+        *ar_object.background_deletion.map { "Association-#{ar_object_str}:#{_1}" },
+      ]
     end
   end
 
-  def destroy_and_delete_associations
-    Array(object.background_deletion).each do |association_config|
-      reflection = BackgroundDeletion::Reflection.new(association_config)
-      next unless destroyable_association?(reflection.name)
+  # @param hierarchy [Array<String>] something like ["Plain-Service-1234", "Association-Service-1234:plans" ...]
+  # @note processes deleting a hierarchy of objects and reschedules itself at current progress after a time limit
+  def perform(*hierarchy)
+    return compatibility(hierarchy) unless hierarchy.first.is_a?(String)
 
-      ReflectionDestroyer.new(object, reflection, caller_worker_hierarchy).destroy_later
+    started = now
+    while now - started < WORK_TIME_LIMIT_SECONDS
+      last_entry = hierarchy.pop
+      return unless last_entry
+
+      Rails.logger.info "Starting background deletion iteration with: #{hierarchy.join(' ')}"
+      hierarchy.concat handle_hierarchy_entry(last_entry)
     end
+
+    self.class.perform_later(*hierarchy) # TODO: assure lock does not affect this scheduling, maybe as a callback
   end
 
-  def destroyable_association?(_association)
-    true
+  # we can just use job.arguments.first but for compatibility mode we want it not to lock
+  def lock_key
+    first_argument = arguments.first
+    first_argument.is_a?(String) ? first_argument : Random.uuid
   end
 
   private
 
-  def called_from_provider_hierarchy?
-    return unless (tenant_id = object.tenant_id)
-    caller_worker_hierarchy.include?("Hierarchy-Account-#{tenant_id}")
+  def hierarchy_entries_for(...)
+    self.class.send(:hierarchy_entries_for, ...)
   end
 
-  def callback_options
-    { 'object_global_id' => object.to_global_id, 'caller_worker_hierarchy' => caller_worker_hierarchy }
+  # handles a single hierarchy entry
+  # @return [Array<String>] hierarchy for a newly discovered object from association to delete or empty array otherwise
+  def handle_hierarchy_entry(entry)
+    case entry
+    when /Plain-([:\w]+)-(\d+)/
+      ar_object = $1.constantize.find($2.to_i)
+      # callbacks logic differs between object destroyed by association, it is standalone if first job arg is itself
+      ar_object.destroyed_by_association = DummyDestroyedByAssociationReflection.new(entry) if arguments.first != entry
+      ar_object.background_deletion_method_call
```

Comment:
There is one difference with previous logic. Previously `#destroy` without `!` was used when destroyed by association. But I don't see why should this be done and so far haven't observed any negative effects. I think it is correct that the job should fail/retry if the `#destroy!` method failed.

Reply:
Oh, OK, this is the response to my previous comment :sweat_smile: 

By the way, do you think `background_deletion_method_call` method is necessary? I think it causes more confusion, and I would rather just do `ar_object.send background_deletion_method` here.

Reply:
it would be something like
```rb
ar_object.public_send(ar.object.background_deletion_method)
```

 which is IMO ugly. Or do I miss something?